## Part 1: Document Preprocessing and Chunking
### Task 1.1: Implement Basic Chunking

Create a function that splits documents into chunks based on a maximum token/word count.

```python
def chunk_by_words(text, max_words=50, overlap=10):
    """
    Split text into chunks based on word count.
    
    Args:
        text: Input text to chunk
        max_words: Maximum words per chunk
        overlap: Number of overlapping words between chunks
    
    Returns:
        List of text chunks
    """
    # Your code here
    pass
```

**Test your implementation:**

```python
sample_text = "Machine learning is a subset of artificial intelligence. It enables systems to learn from experience. Deep learning uses neural networks with multiple layers."

chunks = chunk_by_words(sample_text, max_words=10, overlap=3)
print(f"Number of chunks: {len(chunks)}")
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}: {chunk}")
```

In [ ]:
def chunk_by_words(text, max_words=50, overlap=10):
# -------------------------------
# Step 1: Validate input values
# -------------------------------
# max_words must be positive
    if max_words <= 0:
        raise ValueError("max_words must be greater than 0")
# overlap must be smaller than max_words
# otherwise the sliding window will not move properly
    if overlap >= max_words:
        raise ValueError("overlap must be smaller than max_words")
# -------------------------------
# Step 2: Split text into words
# -------------------------------
# Split text into a list of words using whitespace
    words = text.split()
# List to store final chunks
    chunks = []
# -------------------------------
# Step 3: Define sliding window
# -------------------------------
# Start index of first chunk
    start = 0
# We subtract overlap so that chunks share some words
    step = max_words - overlap
# -------------------------------
# Step 4: Create chunks
# -------------------------------
# Continue until we reach the end of word list
    while start < len(words):
# Define end of current chunk
        end = start + max_words
# Slice words from start to end
        chunk = words[start:end]
# Join words back into a string
        chunks.append(" ".join(chunk))
# Move window forward
        start += step
# -------------------------------
# Step 5: Return all chunks
# -------------------------------
    return chunks
# ==========================================
# TEST THE FUNCTION
# ==========================================
sample_text = (
    "Machine learning is a subset of artificial intelligence. "
    "It enables systems to learn from experience. "
    "Deep learning uses neural networks with multiple layers."
)
# Create chunks with 10 words per chunk
# and 3 words overlapping between chunks
chunks = chunk_by_words(sample_text, max_words=10, overlap=3)
# Print number of chunks
print(f"Number of chunks: {len(chunks)}\n")
# Print each chunk
for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}: {chunk}")

Number of chunks: 4

Chunk 1: Machine learning is a subset of artificial intelligence. It enables
Chunk 2: intelligence. It enables systems to learn from experience. Deep learning
Chunk 3: experience. Deep learning uses neural networks with multiple layers.
Chunk 4: multiple layers.


### Task 1.2: Sentence-Based Chunking

Implement a more sophisticated chunking strategy that respects sentence boundaries.

```python
def chunk_by_sentences(text, max_sentences=2):
    """
    Split text into chunks based on complete sentences.
    
    Args:
        text: Input text to chunk
        max_sentences: Maximum sentences per chunk
    
    Returns:
        List of text chunks
    """
    # Your code here
    # Hint: Use simple sentence splitting on periods followed by spaces
    pass
```

In [ ]:
def chunk_by_sentences(text, max_sentences=2):
# ------------------------------------
# Step 1: Validate input
# ------------------------------------
    if max_sentences <= 0:
        raise ValueError("max_sentences must be greater than 0")
# ------------------------------------
# Step 2: Split text into sentences
# ------------------------------------
# Simple sentence splitting based on ". "
# This assumes sentences end with a period followed by a space
    sentences = text.split(". ")
# remove extra whitespace and ignore empty strings
    sentences = [s.strip() for s in sentences if s.strip()]
# ------------------------------------
# Step 3: Group sentences into chunks
# ------------------------------------
    chunks = []
# Loop through sentences in steps of max_sentences
    for i in range(0, len(sentences), max_sentences):
# Select group of sentences
        chunk_sentences = sentences[i:i + max_sentences]
# Join them back into a paragraph
        chunk = ". ".join(chunk_sentences)
# Ensure the chunk ends with a period
        if not chunk.endswith("."):
            chunk += "."

        chunks.append(chunk)
# ------------------------------------
# Step 4: Return final chunks
# ------------------------------------
    return chunks
# ==========================================
# TEST THE FUNCTION
# ==========================================

sample_text = (
    "Machine learning is a subset of artificial intelligence. "
    "It enables systems to learn from experience. "
    "Deep learning uses neural networks with multiple layers. "
    "Neural networks are inspired by biological systems."
)

chunks = chunk_by_sentences(sample_text, max_sentences=2)

print(f"Number of chunks: {len(chunks)}\n")

for i, chunk in enumerate(chunks):
    print(f"Chunk {i+1}: {chunk}")

Number of chunks: 2

Chunk 1: Machine learning is a subset of artificial intelligence. It enables systems to learn from experience.
Chunk 2: Deep learning uses neural networks with multiple layers. Neural networks are inspired by biological systems.


**Questions to consider:**
- What are the advantages of sentence-based chunking over word-based chunking?

Answer.Sentence-based chunking preserves semantic meaning and produces better embeddings because it keeps complete thoughts together.
- When might word-based chunking be more appropriate?

Answer.Word-based chunking is more useful when  precise chunk sizes are required, especially in large-scale systems.

## Part 2: Creating Embeddings
### Task 2.1: Initialize Embedding Model

Load a pre-trained sentence transformer model.



In [9]:
import numpy as np
from sentence_transformers import SentenceTransformer

# Sample knowledge base (your documents)
DOCUMENTS = [
    "Machine learning is a subset of artificial intelligence.",
    "Deep learning uses neural networks with multiple layers.",
    "Natural language processing enables computers to understand text.",
    "Embeddings convert text into numerical vectors.",
    "Retrieval systems use embeddings to find similar documents."
]

def create_embeddings(documents, model):
    """
    Generate embeddings for a list of documents.a

    Args:
        documents (list): List of text documents
        model: Embedding model

    Returns:
        numpy.ndarray: Array of embeddings
    """
    embeddings = model.encode(documents)
    return np.array(embeddings)


# Load a pre-trained sentence transformer model
model = SentenceTransformer('all-MiniLM-L6-v2')

# Generate embeddings
embeddings = create_embeddings(DOCUMENTS, model)

print("Embeddings created successfully")
print(f"Number of documents: {len(DOCUMENTS)}")
print(f"Embedding shape: {embeddings.shape}")
print(f"First embedding (first 10 dimensions):\n{embeddings[0][:10]}")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Embeddings created successfully
Number of documents: 5
Embedding shape: (5, 384)
First embedding (first 10 dimensions):
[-0.02345927 -0.01058112  0.07192818  0.03042551  0.02855948 -0.03882562
 -0.02819778 -0.01625061 -0.05763257 -0.00336628]


In [ ]:
sample_text = "Machine learning enables computers to learn from data."

embedding = model.encode(sample_text)

print("Embedding created ")
print("Shape:", embedding.shape)
print("First 10 values:", embedding[:10])

Embedding created 
Shape: (384,)
First 10 values: [-0.01599948  0.01208692  0.08066808  0.01379568  0.02841323 -0.0147575
 -0.0363927  -0.09956757  0.01908827  0.00763971]


### Task 2.2: Generate Document Embeddings

Create embeddings for all documents in your knowledge base.


In [ ]:
import numpy as np
# Sample knowledge base (your documents)
DOCUMENTS = [
    "Machine learning is a subset of artificial intelligence.",
    "Deep learning uses neural networks with multiple layers.",
    "Natural language processing enables computers to understand text.",
    "Embeddings convert text into numerical vectors.",
    "Retrieval systems use embeddings to find similar documents."
]
def create_embeddings(documents, model):
    embeddings = model.encode(documents)
    return np.array(embeddings)
# Generate embeddings
embeddings = create_embeddings(DOCUMENTS, model)
print(" Embeddings created successfully")
print(f"Number of documents: {len(DOCUMENTS)}")
print(f"Embedding shape: {embeddings.shape}")
print(f"First embedding (first 10 dimensions):\n{embeddings[0][:10]}")

 Embeddings created successfully
Number of documents: 5
Embedding shape: (5, 384)
First embedding (first 10 dimensions):
[-0.02345927 -0.01058112  0.07192818  0.03042551  0.02855948 -0.03882562
 -0.02819778 -0.01625061 -0.05763257 -0.00336628]



Task 2.3: Understanding Embedding Space
Calculate and visualize the similarity between different documents.

In [10]:
# Import libraries
import numpy as np
from sentence_transformers import SentenceTransformer
# Sample Knowledge Base
DOCUMENTS = [
    "Machine learning is a method where computers learn from data.",
    "Deep learning is a subset of machine learning that uses neural networks.",
    "Natural language processing helps computers understand human language.",
    "Computer vision enables machines to interpret images and videos.",
    "Reinforcement learning is a type of machine learning based on rewards."
]
# Initialize Model
print("Loading embedding model...")
model = SentenceTransformer('all-MiniLM-L6-v2')
# Create Embeddings
print("Generating embeddings...")
embeddings = model.encode(DOCUMENTS)
# Cosine Similarity Function
def cosine_similarity(vec1, vec2):
    """
    Calculate cosine similarity between two vectors.

    Args:
        vec1: First vector
        vec2: Second vector

    Returns:
        Similarity score between -1 and 1
    """
    dot_product = np.dot(vec1, vec2)
    norm_vec1 = np.linalg.norm(vec1)
    norm_vec2 = np.linalg.norm(vec2)
    return dot_product / (norm_vec1 * norm_vec2)
# Compare Similarities
# Document 0: Machine Learning
# Document 1: Deep Learning
# Document 4: Reinforcement Learning

similarity_ml_dl = cosine_similarity(embeddings[0], embeddings[1])
similarity_ml_rl = cosine_similarity(embeddings[0], embeddings[4])

print("\n Similarity Results:")
print(f"Similarity (ML vs DL): {similarity_ml_dl:.4f}")
print(f"Similarity (ML vs RL): {similarity_ml_rl:.4f}")

Loading embedding model...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Generating embeddings...

 Similarity Results:
Similarity (ML vs DL): 0.5923
Similarity (ML vs RL): 0.6346


**Questions:**
- Which pair is more similar? Why does this make sense?

Answer. Machine Learning vs Reinforcement Learning is more similar (0.6346).

- What does the similarity score tell you about semantic relationships?

Answer.Cosine similarity measures how close two vectors are in embedding space.


### Task 3.1: Implement Basic Retrieval

Create a function that retrieves the top-k most similar documents for a given query.


**Test your retrieval system:**

```python
query = "How do computers learn from data?"
results = retrieve_documents(query, DOCUMENTS, embeddings, model, top_k=3)

print(f"Query: {query}\n")
for i, (doc, score) in enumerate(results, 1):
    print(f"Result {i} (Score: {score :.4f}):").
      print(f"{doc}\n")

In [ ]:
# Import Libraries
import numpy as np
from sentence_transformers import SentenceTransformer
import warnings
import os
# Suppress warnings and progress bars
warnings.filterwarnings("ignore")
os.environ["TRANSFORMERS_VERBOSITY"] = "error"
os.environ["HF_HUB_DISABLE_PROGRESS_BARS"] = "1"
# Sample Knowledge Base
DOCUMENTS = [
    "Machine learning is a method where computers learn from data.",
    "Deep learning is a subset of machine learning that uses neural networks.",
    "Natural language processing helps computers understand human language.",
    "Computer vision enables machines to interpret images and videos.",
    "Reinforcement learning is a type of machine learning based on rewards."
]
# Initialize Embedding Model
print("Loading model...")
model = SentenceTransformer('all-MiniLM-L6-v2')
# Create Document Embeddings
print("Generating document embeddings...")
embeddings = model.encode(DOCUMENTS)
# Cosine Similarity Function
def cosine_similarity(vec1, vec2):
    return np.dot(vec1, vec2) / (
        np.linalg.norm(vec1) * np.linalg.norm(vec2)
    )
# Retrieval Function
def retrieve_documents(query, documents, embeddings, model, top_k=3):
    # Encode query
    query_embedding = model.encode(query)

    # Compute similarity with each document
    similarities = []

    for i, doc_embedding in enumerate(embeddings):
        score = cosine_similarity(query_embedding, doc_embedding)
        similarities.append((documents[i], score))

    #Sort by similarity (highest first)
    similarities.sort(key=lambda x: x[1], reverse=True)

    #Return top_k results
    return similarities[:top_k]
# Test Retrieval
query = "How do computers learn from data?"
results = retrieve_documents(query, DOCUMENTS, embeddings, model, top_k=3)
print("\nQuery:", query, "\n")
for i, (doc, score) in enumerate(results, 1):
    print(f"Result {i} (Score: {score:.4f}):")
    print(doc)
    print()

Loading model...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Generating document embeddings...

Query: How do computers learn from data? 

Result 1 (Score: 0.6801):
Machine learning is a method where computers learn from data.

Result 2 (Score: 0.4494):
Computer vision enables machines to interpret images and videos.

Result 3 (Score: 0.4417):
Natural language processing helps computers understand human language.



**Questions:**
1.What happens when you query about topics not in your knowledge base?

Answer. The system produces low similarity scores and may return irrelevant documents.

2.How does the threshold help filter irrelevant results?

Answer.The threshold removes documents with low similarity scores, ensuring only sufficiently relevant results are returned.



### Task 4.1: Implement Batch Processing

Create a function that efficiently handles multiple queries at once.


In [2]:
# COMPLETE BATCH RETRIEVAL SYSTEM
import numpy as np
from sentence_transformers import SentenceTransformer
import warnings
import os
# Suppress warnings
warnings.filterwarnings("ignore")
os.environ["TRANSFORMERS_VERBOSITY"] = "error"
os.environ["HF_HUB_DISABLE_PROGRESS_BARS"] = "1"
# Knowledge Base
# --------------
DOCUMENTS = [
    "Machine learning is a method where computers learn from data.",
    "Deep learning is a subset of machine learning that uses neural networks.",
    "Natural language processing helps computers understand human language.",
    "Computer vision enables machines to interpret images and videos.",
    "Reinforcement learning is a type of machine learning based on rewards."
]
# Load Embedding Model
print("Loading model...")
model = SentenceTransformer('all-MiniLM-L6-v2')
# Create Document Embeddings
print("Generating document embeddings...")
embeddings = model.encode(DOCUMENTS)
# Batch Retrieval Function
def batch_retrieve(queries, documents, embeddings, model, top_k=3):
    """
    Efficiently retrieve documents for multiple queries.
    """
    results_dict = {}

    # Create embeddings for all queries at once
    query_embeddings = model.encode(queries)

    # Normalize document embeddings
    doc_norms = np.linalg.norm(embeddings, axis=1)

    for query, query_embedding in zip(queries, query_embeddings):

        query_norm = np.linalg.norm(query_embedding)

        # Compute cosine similarity (vectorized)
        similarities = np.dot(embeddings, query_embedding) / (
            doc_norms * query_norm
        )

        # Get top_k indices
        top_indices = np.argsort(similarities)[::-1][:top_k]

        # Store results
        results = [(documents[i], similarities[i]) for i in top_indices]
        results_dict[query] = results

    return results_dict
# Testing  Batch Retrieval
queries = [
    "What is machine learning?",
    "Explain neural networks",
    "Tell me about reinforcement learning",
    "Tell me about quantum computing"  # Out-of-domain
]

batch_results = batch_retrieve(queries, DOCUMENTS, embeddings, model, top_k=2)
# Display Results
for query, results in batch_results.items():
    print("\nQuery:", query)

    for i, (doc, score) in enumerate(results, 1):
        print(f"Result {i} (Score: {score:.4f}):")
        print(doc)
        print()

Loading model...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Generating document embeddings...

Query: What is machine learning?
Result 1 (Score: 0.8877):
Machine learning is a method where computers learn from data.

Result 2 (Score: 0.6171):
Deep learning is a subset of machine learning that uses neural networks.


Query: Explain neural networks
Result 1 (Score: 0.5547):
Deep learning is a subset of machine learning that uses neural networks.

Result 2 (Score: 0.4547):
Computer vision enables machines to interpret images and videos.


Query: Tell me about reinforcement learning
Result 1 (Score: 0.8025):
Reinforcement learning is a type of machine learning based on rewards.

Result 2 (Score: 0.4928):
Machine learning is a method where computers learn from data.


Query: Tell me about quantum computing
Result 1 (Score: 0.2768):
Reinforcement learning is a type of machine learning based on rewards.

Result 2 (Score: 0.2643):
Deep learning is a subset of machine learning that uses neural networks.



### Task 4.2: Create a Retrieval Class

Encapsulate your retrieval pipeline in a reusable class.



**Test your class:**

```python
from documents import DOCUMENTS

pipeline = RetrievalPipeline()
pipeline.add_documents(DOCUMENTS)

print("Pipeline Statistics:")
print(pipeline.get_stats())

query = "What are neural networks?"
results = pipeline.retrieve(query, top_k=2)

print(f"\nQuery: {query}")
for doc, score in results:
    print(f"Score: {score:.4f} | {doc[:80]}...")
```

In [8]:
# DEFINE DOCUMENTS (Knowledge Base)
DOCUMENTS = [
    "Machine learning is a method where computers learn from data.",
    "Deep learning is a subset of machine learning that uses neural networks.",
    "Natural language processing helps computers understand human language.",
    "Computer vision enables machines to interpret images and videos.",
    "Reinforcement learning is a type of machine learning based on rewards."
]
# IMPORT REQUIRED LIBRARIES
import numpy as np
from sentence_transformers import SentenceTransformer
import warnings
import os
# Suppress unnecessary warnings and progress bars
warnings.filterwarnings("ignore")
os.environ["TRANSFORMERS_VERBOSITY"] = "error"
os.environ["HF_HUB_DISABLE_PROGRESS_BARS"] = "1"
# RETRIEVAL PIPELINE CLASS
class RetrievalPipeline:
    # Constructor → Loads embedding model
    def __init__(self, model_name='all-MiniLM-L6-v2'):
        print("Initializing RetrievalPipeline...")
        self.model = SentenceTransformer(model_name)  # Load embedding model
        self.documents = []                          # Store documents
        self.embeddings = None                       # Store document embeddings
        print("Model loaded.")

    # Add documents and generate embeddings
    def add_documents(self, documents):
        print(f"Adding {len(documents)} documents...")
        self.documents.extend(documents)             # Store documents
        self._generate_embeddings()                  # Create embeddings
        print("Documents added and embeddings generated.")

    # Private method → Generate embeddings for documents
    def _generate_embeddings(self):
        if self.documents:
            self.embeddings = self.model.encode(self.documents)
        else:
            self.embeddings = None

    # Private method → Compute cosine similarity
    def _cosine_similarity(self, vec1, vec2):
        return np.dot(vec1, vec2) / (
            np.linalg.norm(vec1) * np.linalg.norm(vec2)
        )

    # Retrieve top_k most similar documents
    def retrieve(self, query, top_k=3):

        # Check if documents exist
        if self.embeddings is None or not self.documents:
            raise ValueError("No documents added. Call add_documents() first.")

        # Convert query into embedding
        query_embedding = self.model.encode(query)

        similarities = []

        # Compute similarity with each document
        for i, doc_embedding in enumerate(self.embeddings):
            score = self._cosine_similarity(query_embedding, doc_embedding)
            similarities.append((self.documents[i], score))

        # Sort by highest similarity
        similarities.sort(key=lambda x: x[1], reverse=True)

        # Return top_k results
        return similarities[:top_k]

    # Return pipeline statistics
    def get_stats(self):
        return {
            "num_documents": len(self.documents),
            "embedding_dimension": self.embeddings.shape[1]
                if self.embeddings is not None else 0
        }
# TEST THE PIPELINE
# Initialize pipeline
pipeline = RetrievalPipeline()
# Add documents to pipeline
pipeline.add_documents(DOCUMENTS)
# Print statistics
print("Pipeline Statistics:")
print(pipeline.get_stats())
# Test query
query = "What are neural networks?"
results = pipeline.retrieve(query, top_k=2)
# Print results
print(f"\nQuery: {query}")
for doc, score in results:
    print(f"Score: {score:.4f} | {doc[:80]}...")

Initializing RetrievalPipeline...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Model loaded.
Adding 5 documents...
Documents added and embeddings generated.
Pipeline Statistics:
{'num_documents': 5, 'embedding_dimension': 384}

Query: What are neural networks?
Score: 0.6360 | Deep learning is a subset of machine learning that uses neural networks....
Score: 0.4394 | Machine learning is a method where computers learn from data....


### Task 5.1: Implement Relevance Metrics

Create functions to measure retrieval quality.

```python
def calculate_precision_at_k(retrieved_docs, relevant_docs, k=3):
    """
    Calculate precision at k: proportion of retrieved docs that are relevant.
    
    Args:
        retrieved_docs: List of retrieved document indices
        relevant_docs: Set of relevant document indices
        k: Number of top results to consider
    
    Returns:
        Precision score
    """
    # Your code here
    pass

def calculate_recall_at_k(retrieved_docs, relevant_docs, k=3):
    """
    Calculate recall at k: proportion of relevant docs that were retrieved.
    
    Args:
        retrieved_docs: List of retrieved document indices
        relevant_docs: Set of relevant document indices
        k: Number of top results to consider
    
    Returns:
        Recall score
    """
    # Your code here
    pass
```

In [14]:
# RELEVANCE METRICS
def calculate_precision_at_k(retrieved_docs, relevant_docs, k=3):
    retrieved_at_k = retrieved_docs[:k]
    relevant_count = sum(1 for doc in retrieved_at_k if doc in relevant_docs)
    return relevant_count / k if k > 0 else 0.0
def calculate_recall_at_k(retrieved_docs, relevant_docs, k=3):
    retrieved_at_k = retrieved_docs[:k]
    relevant_count = sum(1 for doc in retrieved_at_k if doc in relevant_docs)
    return relevant_count / len(relevant_docs) if len(relevant_docs) > 0 else 0.0
# Testing examples
retrieved_docs = [1, 0, 4]   # Example retrieved indices
relevant_docs = {0, 1}       # Ground truth relevant docs

precision = calculate_precision_at_k(retrieved_docs, relevant_docs, k=3)
recall = calculate_recall_at_k(retrieved_docs, relevant_docs, k=3)

print("Retrieved Docs:", retrieved_docs)
print("Relevant Docs:", relevant_docs)
print(f"Precision@3: {precision:.4f}")
print(f"Recall@3: {recall:.4f}")

Retrieved Docs: [1, 0, 4]
Relevant Docs: {0, 1}
Precision@3: 0.6667
Recall@3: 1.0000


In [17]:
# TEST CASES
# (query, set of relevant document indices)
test_cases = [
    ("What is machine learning?", {0}),
    ("Explain neural networks", {1}),
    ("How does computer vision work?", {3}),
]
# EVALUATION FUNCTION
def evaluate_retrieval(pipeline, test_cases, k=3):
    total_precision = 0
    total_recall = 0
    for query, relevant_docs in test_cases:
# Retrieve documents
        results = pipeline.retrieve(query, top_k=k)
# Get indices of retrieved documents
        retrieved_indices = []
        for doc, _ in results:
            retrieved_indices.append(pipeline.documents.index(doc))
# Calculate precision and recall
        precision = calculate_precision_at_k(retrieved_indices, relevant_docs, k)
        recall = calculate_recall_at_k(retrieved_indices, relevant_docs, k)

        total_precision += precision
        total_recall += recall
#  Average across all test cases
    avg_precision = total_precision / len(test_cases)
    avg_recall = total_recall / len(test_cases)

    return {
        "average_precision_at_k": avg_precision,
        "average_recall_at_k": avg_recall
    }
# Run evaluation
metrics = evaluate_retrieval(pipeline, test_cases, k=3)
print("Evaluation Results:")
print(f"Average Precision@3: {metrics['average_precision_at_k']:.4f}")
print(f"Average Recall@3: {metrics['average_recall_at_k']:.4f}")

Evaluation Results:
Average Precision@3: 0.3333
Average Recall@3: 1.0000


### Task 5.3: Analyze Retrieval Performance

Run evaluation and analyze the results.

```python
# Your code to evaluate and print results
```

In [19]:
# TASK 5.3 – ANALYZE RETRIEVAL PERFORMANCE
# Run overall evaluation (average precision & recall)
metrics = evaluate_retrieval(pipeline, test_cases, k=3)
print("===== Overall Evaluation Results =====")
# Print average precision across all test queries
print(f"Average Precision@3: {metrics['average_precision_at_k']:.4f}")
# Print average recall across all test queries
print(f"Average Recall@3: {metrics['average_recall_at_k']:.4f}")
print("\n===== Per Query Detailed Analysis =====")
#Analyze performance for each individual query
for query, relevant_docs in test_cases:
    # Retrieve top 3 documents for the query
    results = pipeline.retrieve(query, top_k=3)
    # Convert retrieved documents into their indices
    # (needed to compare with relevant document indices)
    retrieved_indices = [
        pipeline.documents.index(doc) for doc, _ in results
    ]
    # Calculate precision for this query
    precision = calculate_precision_at_k(
        retrieved_indices, relevant_docs, k=3
    )
    # Calculate recall for this query
    recall = calculate_recall_at_k(
        retrieved_indices, relevant_docs, k=3
    )
    # Print detailed results
    print(f"\nQuery: {query}")
    print(f"Relevant Document Indices: {relevant_docs}")
    print(f"Retrieved Document Indices: {retrieved_indices}")
    print(f"Precision@3: {precision:.4f}")
    print(f"Recall@3: {recall:.4f}")

===== Overall Evaluation Results =====
Average Precision@3: 0.3333
Average Recall@3: 1.0000

===== Per Query Detailed Analysis =====

Query: What is machine learning?
Relevant Document Indices: {0}
Retrieved Document Indices: [0, 1, 4]
Precision@3: 0.3333
Recall@3: 1.0000

Query: Explain neural networks
Relevant Document Indices: {1}
Retrieved Document Indices: [1, 3, 0]
Precision@3: 0.3333
Recall@3: 1.0000

Query: How does computer vision work?
Relevant Document Indices: {3}
Retrieved Document Indices: [3, 2, 0]
Precision@3: 0.3333
Recall@3: 1.0000


1.What is the average precision and recall of your system?

Answer.The system shows moderate to high average precision and recall, meaning most relevant documents are retrieved within the top results.

2️.Which queries perform best/worst? Why?

Answer.Queries closely matching document content perform best, while broader or less specific queries perform worse due to semantic overlap.

3.How could you improve retrieval quality?

Answer.Retrieval quality can be improved by using better embedding models, adding more relevant data, and applying re-ranking techniques.

In [ ]:
# ==========================================================
# TASK 6.1 – EXTENDED KNOWLEDGE BASE
# ==========================================================

EXTENDED_DOCS = [

    """
    Artificial Intelligence (AI) is transforming industries worldwide. Machine learning,
    a subset of AI, enables systems to learn from data without explicit programming.
    Companies use ML for predictive analytics, recommendation systems, fraud detection,
    and intelligent automation. The rapid growth of big data and cloud computing has
    accelerated AI adoption across healthcare, finance, retail, and manufacturing sectors.
    """,

    """
    Deep learning has revolutionized computer vision and natural language processing.
    Convolutional Neural Networks (CNNs) excel at image classification and object detection,
    while Recurrent Neural Networks (RNNs) and Transformers handle sequential data.
    Applications include medical image diagnosis, autonomous vehicles, speech recognition,
    and biometric security systems.
    """,

    """
    Natural Language Processing (NLP) enables machines to understand, interpret, and
    generate human language. Modern NLP models use transformer architectures such as
    BERT and GPT to achieve state-of-the-art performance. These systems power chatbots,
    translation services, sentiment analysis tools, and intelligent search engines.
    NLP plays a critical role in human-computer interaction.
    """,

    """
    Reinforcement learning is a machine learning paradigm where agents learn by interacting
    with an environment. The agent receives rewards or penalties based on its actions
    and optimizes its strategy over time. Reinforcement learning is widely used in
    robotics, game playing (such as AlphaGo), recommendation systems, and resource optimization.
    """,

    """
    Computer vision focuses on enabling machines to interpret and analyze visual data
    from the world. Techniques include image segmentation, object detection, facial
    recognition, and motion tracking. Computer vision systems are used in surveillance,
    healthcare diagnostics, smart cities, and augmented reality applications.
    """,

    """
    Data science combines statistics, programming, and domain expertise to extract
    meaningful insights from structured and unstructured data. Data scientists use
    tools like Python, R, SQL, and visualization libraries to analyze trends and
    build predictive models. Data-driven decision making has become essential for
    modern organizations.
    """
]


### Task 6.2: Handle Multi-Document Queries

Implement a function that retrieves and combines information from multiple documents.


In [20]:
# TASK 6.2 – MULTI-DOCUMENT RETRIEVAL
def multi_document_retrieve(query, pipeline, top_k=5, combine=True):
    #Retrieve top_k documents using pipeline
    results = pipeline.retrieve(query, top_k=top_k)
 #If combine=False → return raw results
    if not combine:
        return results
# combine retrieved documents into one context string
    combined_context = ""
    for i, (doc, score) in enumerate(results, 1):
        combined_context += f"\n--- Document {i} (Score: {score:.4f}) ---\n"
        combined_context += doc.strip() + "\n"

    return combined_context
# TEST MULTI-DOCUMENT RETRIEVAL
query = "How is deep learning used in computer vision and NLP?"
combined_result = multi_document_retrieve(
    query,
    pipeline,
    top_k=3,
    combine=True
)

print("Query:", query)
print("\nCombined Context:")
print(combined_result)

Query: How is deep learning used in computer vision and NLP?

Combined Context:

--- Document 1 (Score: 0.6756) ---
Deep learning is a subset of machine learning that uses neural networks.

--- Document 2 (Score: 0.5329) ---
Computer vision enables machines to interpret images and videos.

--- Document 3 (Score: 0.4567) ---
Natural language processing helps computers understand human language.



### Task 6.3: Build a Complete RAG Retrieval Demo

Create a simple interactive demo of your retrieval system.



In [23]:
# TASK 6.3 – INTERACTIVE RAG RETRIEVAL DEMO
def interactive_retrieval_demo(pipeline):
    print("RAG Retrieval System Demo")
    print("=" * 50)
    print("Enter your queries (type 'quit' to exit)")
    print("=" * 50)
    while True:
        # Take user input
        query = input("\nEnter Query: ")
        # Exit condition
        if query.lower() == "quit":
            print("\nExiting demo. Goodbye!")
            break
        try:
            #Retrieve top 3 documents
            results = pipeline.retrieve(query, top_k=3)
            print("\nTop Retrieved Documents:")
            print("-" * 50)
            for i, (doc, score) in enumerate(results, 1):
                print(f"\nResult {i}")
                print(f"Similarity Score: {score:.4f}")
                print(f"Document Preview: {doc.strip()[:200]}...")
            #Show system statistics
            stats = pipeline.get_stats()
            print("\nSystem Statistics:")
            print("-" * 50)
            print(f"Total Documents: {stats['num_documents']}")
            print(f"Embedding Dimension: {stats['embedding_dimension']}")
        except Exception as e:
            print("Error:", str(e))
# RUN THE DEMO
interactive_retrieval_demo(pipeline)

RAG Retrieval System Demo
Enter your queries (type 'quit' to exit)

Enter Query: How does reinforcement learning work?

Top Retrieved Documents:
--------------------------------------------------

Result 1
Similarity Score: 0.7612
Document Preview: Reinforcement learning is a type of machine learning based on rewards....

Result 2
Similarity Score: 0.4993
Document Preview: Machine learning is a method where computers learn from data....

Result 3
Similarity Score: 0.3198
Document Preview: Deep learning is a subset of machine learning that uses neural networks....

System Statistics:
--------------------------------------------------
Total Documents: 5
Embedding Dimension: 384

Enter Query: quit

Exiting demo. Goodbye!


Challenge Tasks

### Challenge 1: Implement Re-ranking

After initial retrieval, re-rank results using a different strategy (e.g., keyword matching combined with semantic similarity).

In [24]:
import numpy as np
import re
# Simple keyword overlap scorer
def keyword_overlap_score(query, document):
    # Convert to lowercase and tokenize
    query_words = set(re.findall(r"\w+", query.lower()))
    doc_words = set(re.findall(r"\w+", document.lower()))
    if len(query_words) == 0:
        return 0.0
    overlap = query_words.intersection(doc_words)
    return len(overlap) / len(query_words)
# Re-ranking function
def retrieve_with_reranking(query, pipeline, top_k=3,
                             semantic_weight=0.7, keyword_weight=0.3):

#Initial semantic retrieval (get more than needed)
    initial_results = pipeline.retrieve(query, top_k=top_k * 2)

    reranked_results = []

    for doc, semantic_score in initial_results:

# Compute keyword overlap
        keyword_score = keyword_overlap_score(query, doc)
#Combine scores
        final_score = (
            semantic_weight * semantic_score +
            keyword_weight * keyword_score
        )

        reranked_results.append((doc, final_score))
#Sort by combined score
    reranked_results.sort(key=lambda x: x[1], reverse=True)

    return reranked_results[:top_k]
# TEST RE-RANKING
query = "Explain neural networks in deep learning"
results = retrieve_with_reranking(query, pipeline, top_k=3)
print("Query:", query)
print("\nRe-ranked Results:")

for i, (doc, score) in enumerate(results, 1):
    print(f"\nResult {i}")
    print(f"Final Score: {score:.4f}")
    print(doc[:150] + "...")

Query: Explain neural networks in deep learning

Re-ranked Results:

Result 1
Final Score: 0.6426
Deep learning is a subset of machine learning that uses neural networks....

Result 2
Final Score: 0.3414
Machine learning is a method where computers learn from data....

Result 3
Final Score: 0.3193
Computer vision enables machines to interpret images and videos....


### Challenge 2: Add Metadata Filtering

Extend your pipeline to support filtering documents by metadata (e.g., date, category, author).


In [28]:
# METADATA-AWARE RAG RETRIEVAL SYSTEM (FULL WORKING CODE)
import numpy as np
from sentence_transformers import SentenceTransformer
#DOCUMENTS WITH METADATA
DOCUMENTS_WITH_METADATA = [
    {
        "text": "Machine learning enables systems to learn from data.",
        "category": "AI",
        "author": "Alice",
        "year": 2022
    },
    {
        "text": "Deep learning uses neural networks for complex tasks.",
        "category": "AI",
        "author": "Bob",
        "year": 2023
    },
    {
        "text": "Computer vision focuses on image recognition.",
        "category": "Vision",
        "author": "Alice",
        "year": 2021
    },
    {
        "text": "Natural language processing enables chatbots.",
        "category": "NLP",
        "author": "Charlie",
        "year": 2023
    }
]
# METADATA RETRIEVAL PIPELINE
class MetadataRetrievalPipeline:

    def __init__(self, model_name='all-MiniLM-L6-v2'):
        print("Initializing model...")
        self.model = SentenceTransformer(model_name)
        self.documents = []
        self.embeddings = None

    def add_documents(self, documents):
        self.documents.extend(documents)
        texts = [doc["text"] for doc in self.documents]
        print("Encoding documents...")
        self.embeddings = self.model.encode(texts)
        print("Documents encoded successfully!\n")
    def _cosine_similarity(self, vec1, vec2):
        return np.dot(vec1, vec2) / (
            np.linalg.norm(vec1) * np.linalg.norm(vec2)
        )
    def retrieve(self, query, top_k=3, filters=None):
        """
        Retrieve documents with optional metadata filtering
        """
        if self.embeddings is None:
            print("No documents added yet!")
            return []
        print(f"\nQuery: {query}")
        print(f"Filters: {filters}\n")
        query_embedding = self.model.encode(query)
        similarities = []
        for i, doc in enumerate(self.documents):
            # Apply Metadata Filtering
            if filters:
                match = True
                for key, value in filters.items():
                    if doc.get(key) != value:
                        match = False
                        break
                if not match:
                    continue
            # Compute Similarity
            score = self._cosine_similarity(
                query_embedding,
                self.embeddings[i]
            )

            similarities.append((doc, score))
        # Sort by score (highest first)
        similarities.sort(key=lambda x: x[1], reverse=True)

        return similarities[:top_k]
#  RUN THE SYSTEM
if __name__ == "__main__":

    # Initialize pipeline
    pipeline = MetadataRetrievalPipeline()

    # Add documents
    pipeline.add_documents(DOCUMENTS_WITH_METADATA)
 # Example 1: No Filter
    results = pipeline.retrieve("neural networks", top_k=3)
    print("Results (No Filter):")
    for doc, score in results:
        print(f"Score: {score:.4f}")
        print("Text:", doc["text"])
        print("Metadata:", {k:v for k,v in doc.items() if k != "text"})
        print("-" * 50)
  # Example 2: Filter by Category
    results = pipeline.retrieve(
        "neural networks",
        top_k=3,
        filters={"category": "AI"}
    )

    print("\nResults (Filter: category = AI):")
    for doc, score in results:
        print(f"Score: {score:.4f}")
        print("Text:", doc["text"])
        print("Metadata:", {k:v for k,v in doc.items() if k != "text"})
        print("-" * 50)
  # Example 3: Filter by Author
    results = pipeline.retrieve(
        "image recognition",
        top_k=3,
        filters={"author": "Alice"}
    )
    print("\nResults (Filter: author = Alice):")
    for doc, score in results:
        print(f"Score: {score:.4f}")
        print("Text:", doc["text"])
        print("Metadata:", {k:v for k,v in doc.items() if k != "text"})
        print("-" * 50)

Initializing model...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Encoding documents...
Documents encoded successfully!


Query: neural networks
Filters: None

Results (No Filter):
Score: 0.5430
Text: Deep learning uses neural networks for complex tasks.
Metadata: {'category': 'AI', 'author': 'Bob', 'year': 2023}
--------------------------------------------------
Score: 0.3965
Text: Machine learning enables systems to learn from data.
Metadata: {'category': 'AI', 'author': 'Alice', 'year': 2022}
--------------------------------------------------
Score: 0.3440
Text: Computer vision focuses on image recognition.
Metadata: {'category': 'Vision', 'author': 'Alice', 'year': 2021}
--------------------------------------------------

Query: neural networks
Filters: {'category': 'AI'}


Results (Filter: category = AI):
Score: 0.5430
Text: Deep learning uses neural networks for complex tasks.
Metadata: {'category': 'AI', 'author': 'Bob', 'year': 2023}
--------------------------------------------------
Score: 0.3965
Text: Machine learning enables systems to lea

### Challenge 3: Implement Hybrid Search

Combine dense retrieval (embeddings) with sparse retrieval (keyword-based like BM25) for improved results.

In [29]:
# HYBRID SEARCH (DENSE + SPARSE)
import numpy as np
import re
from collections import Counter
from sentence_transformers import SentenceTransformer
# SAMPLE DOCUMENTS
DOCUMENTS = [
    "Machine learning enables systems to learn from data.",
    "Deep learning uses neural networks for complex tasks.",
    "Computer vision focuses on image recognition.",
    "Natural language processing enables chatbots.",
    "Transformers like BERT and GPT are used in NLP."
]
# ----------------------------------------------------------
# HYBRID RETRIEVAL PIPELINE
# ----------------------------------------------------------

class HybridRetrievalPipeline:

    def __init__(self, model_name='all-MiniLM-L6-v2'):
        print("Initializing Hybrid Retrieval System...")
        self.model = SentenceTransformer(model_name)
        self.documents = []
        self.embeddings = None
        self.tokenized_docs = []

    # ------------------------------------------------------
    # Add Documents
    # ------------------------------------------------------

    def add_documents(self, documents):
        self.documents = documents

        # Dense embeddings
        print("Encoding documents (Dense)...")
        self.embeddings = self.model.encode(documents)

        # Sparse tokenization
        print("Tokenizing documents (Sparse)...")
        self.tokenized_docs = [
            re.findall(r"\w+", doc.lower()) for doc in documents
        ]

        print("System ready!\n")
    # ------------------------------------------------------
    # Cosine Similarity (Dense)
    # ------------------------------------------------------

    def _cosine_similarity(self, vec1, vec2):
        return np.dot(vec1, vec2) / (
            np.linalg.norm(vec1) * np.linalg.norm(vec2)
        )
    # ------------------------------------------------------
    # Sparse Keyword Score (Simple TF matching)
    # ------------------------------------------------------

    def _keyword_score(self, query_tokens, doc_tokens):
        doc_counter = Counter(doc_tokens)
        score = 0

        for token in query_tokens:
            score += doc_counter[token]

        return score
    # ------------------------------------------------------
    # Hybrid Retrieve
    # ------------------------------------------------------

    def retrieve(self, query, top_k=3, alpha=0.7):
        """
        alpha → weight for dense score
        (1 - alpha) → weight for sparse score
        """

        query_embedding = self.model.encode(query)
        query_tokens = re.findall(r"\w+", query.lower())

        results = []

        for i, doc in enumerate(self.documents):

            # Dense Score
            dense_score = self._cosine_similarity(
                query_embedding,
                self.embeddings[i]
            )

            # Sparse Score
            sparse_score = self._keyword_score(
                query_tokens,
                self.tokenized_docs[i]
            )

            # Normalize sparse score
            sparse_score = sparse_score / (len(self.tokenized_docs[i]) + 1)

            # Hybrid Score
            hybrid_score = alpha * dense_score + (1 - alpha) * sparse_score

            results.append((doc, hybrid_score, dense_score, sparse_score))

        # Sort by hybrid score
        results.sort(key=lambda x: x[1], reverse=True)

        return results[:top_k]
# ----------------------------------------------------------
# RUN HYBRID SEARCH
# ----------------------------------------------------------
if __name__ == "__main__":

    pipeline = HybridRetrievalPipeline()
    pipeline.add_documents(DOCUMENTS)

    query = "Explain neural networks in deep learning"

    results = pipeline.retrieve(query, top_k=3, alpha=0.7)

    print("Query:", query)
    print("=" * 60)

    for doc, hybrid, dense, sparse in results:
        print(f"Hybrid Score: {hybrid:.4f}")
        print(f"Dense Score : {dense:.4f}")
        print(f"Sparse Score: {sparse:.4f}")
        print("Document:", doc)
        print("-" * 60)

Initializing Hybrid Retrieval System...


Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


Encoding documents (Dense)...
Tokenizing documents (Sparse)...
System ready!

Query: Explain neural networks in deep learning
Hybrid Score: 0.5772
Dense Score : 0.6341
Sparse Score: 0.4444
Document: Deep learning uses neural networks for complex tasks.
------------------------------------------------------------
Hybrid Score: 0.3177
Dense Score : 0.4062
Sparse Score: 0.1111
Document: Machine learning enables systems to learn from data.
------------------------------------------------------------
Hybrid Score: 0.2389
Dense Score : 0.3413
Sparse Score: 0.0000
Document: Computer vision focuses on image recognition.
------------------------------------------------------------


 Questions

1.What are the key trade-offs between chunk size and retrieval quality?

Answer. Smaller chunks improve precision but may lose context, while larger chunks keep context but can add noise. A balanced chunk size gives better overall results.

2.How would you handle documents in multiple languages?

Answer. Use multilingual embedding models and language detection. Optionally, translate all content into a common language for consistency.

3.What strategies could you use to update your knowledge base without recomputing all embeddings?

Answer.Only generate embeddings for new or updated documents instead of recomputing everything. Use incremental indexing in a vector database.

4.How would you monitor retrieval quality in a production system?

Answer.Track metrics like precision and recall, and monitor user behavior such as clicks and query reformulations.

5.What are the computational costs of your pipeline, and how would you optimize for large-scale deployment?

Answer.Embedding generation and similarity search are expensive. Use ANN search, batching, caching, and scalable vector databases to optimize.
